### PR TITLE
Simplifying the environment setup

### PR DESCRIPTION
**Pull request by TarekkMA**
Use cross to simplify the process of cross compiling 

Related to #9

+ Added a flag to disable formatting using `rustfmt`
+ Changed the anki submodule to use my fork where I added the same flag
+ Moved anki to be under `rslib-bridge` because of a limitation using cross with local dependencies.
+ Modified gradle to utilize cross

### CODE REVIEW

**Comment by TarekkMA** [comment]
Added a `doctor.sh`. To check if environment is set up.

File: .gitmodules [hunk 1]
```diff
@@ -1,4 +1,4 @@
 [submodule "anki"]
-	path = anki
-	url = https://github.com/david-allison-1/anki
-	branch = ankidroid-2-1-34
+	path = rslib-bridge/anki
+	url = https://github.com/TarekkMA/anki
+	branch = ankidriod-2-1-34-1
```
Comment (david-allison-1) [inline .gitmodules:4]: Minor (no need to fix): typo

File: tools/doctor.sh [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+
+if [[ $(which docker) && $(docker --version) ]]; then
+  ok_echo "Docker is installed"
+else
+  error_echo "Docker is not installed. please install it first"
+fi
+
+
+cecho $lgray "Building docker images"
```
Comment (david-allison-1) [inline tools/doctor.sh:46]: This is a fantastic start!

It would be useful to highlight the manual steps which are required for now, so we can eventually automate the entire setup process.

File: tools/doctor.sh [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+./tools/build-docker-images.sh
+
```
Comment (david-allison-1) [inline tools/doctor.sh:48]: It would be an aim to eventually have our CI point to this file

File: tools/doctor.sh [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+cecho $lgray "##################################\n"
+
+cecho $lgray "Getting all submodules recursively"
+git submodule update --init --recursive
+
+
+if [[ $(which docker) && $(docker --version) ]]; then
+  ok_echo "Docker is installed"
+else
+  error_echo "Docker is not installed. please install it first"
```
Comment (david-allison-1) [inline tools/doctor.sh:42]: For our error messages, if we can't set up the environment, we should give specific instructions as to what's needed

File: rslib-bridge/Dockerfiles/Dockerfile.linux [hunk 5]
```diff
@@ -0,0 +1,4 @@
+FROM rustembedded/cross:x86_64-unknown-linux-gnu
```
Comment (david-allison-1) [inline rslib-bridge/Dockerfiles/Dockerfile.linux:1]: Security: can we pin the images

File: docs/ENVIROMENT.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# Enviroment
```
Comment (david-allison-1) [inline docs/ENVIROMENT.md:1]:
```suggestion
# Environment
```

File: rslib-bridge/Cross.toml [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    "RUST_DEBUG",
+    "DONT_RUSTFMT"
+]
+
+volumes = [
+    "OUT_DIR",
+]
+
+
+[target.x86_64-unknown-linux-gnu]
```
Comment (david-allison-1) [inline rslib-bridge/Cross.toml:14]: (maybe too simple - implementer's choice) - possibly explain where these images are generated

**Review by david-allison-1 — COMMENTED** [review]
This is fantastic! - I'll defer to @mikehardy for a more thorough review - he's much stronger on Unix/CI than I am

File: .gitmodules (hunk 1 above)
Comment (TarekkMA) [reply .gitmodules:4]: 😅  oh I always get "andriod" it wrong.

File: tools/doctor.sh (hunk 3 above)
Comment (TarekkMA) [reply tools/doctor.sh:48]: Yes, but we should beef up the doctor script a little more, right now it's only pointing to what's wrong. I did what I could with my small bash knowlege.

File: rslib-bridge/Cross.toml (hunk 7 above)
Comment (TarekkMA) [reply rslib-bridge/Cross.toml:14]: I wish we can build those images here using CI and passing them to docker hub, so users can just download the images.

File: tools/doctor.sh (hunk 2 above)
Comment (TarekkMA) [reply tools/doctor.sh:46]: You mean we should highlight the manual steps in the script? If so we can link to the ENVIRONMENT.md file

Comment (david-allison-1) [reply tools/doctor.sh:46]: I think `TODO` comments in the file (maybe linking to GitHub issues?), so they're not going to go out of date

File: docs/ENVIROMENT.md (hunk 6 above)
Comment (TarekkMA) [reply docs/ENVIROMENT.md]: Done

File: rslib-bridge/Dockerfiles/Dockerfile.linux (hunk 5 above)
Comment (TarekkMA) [reply rslib-bridge/Dockerfiles/Dockerfile.linux:1]: mmm I don't know how we can achieve that, they most recent image with a version is 9 month old

https://hub.docker.com/r/rustembedded/cross/tags?page=1&ordering=last_updated&name=x86_64-unknown-linux-gnu

Comment (david-allison-1) [reply rslib-bridge/Dockerfiles/Dockerfile.linux:1]: (uninformed) Can't we pin via the SHA?

https://rockbag.medium.com/why-you-should-pin-your-docker-images-with-sha-instead-of-tags-fd132443b8a6

**Comment by david-allison-1** [comment]
CI Failure: https://github.com/david-allison-1/Anki-Android-Backend/pull/39/checks?check_run_id=2187896650

```
bash: cross: command not found
```

**Comment by TarekkMA** [comment]
Should we push the docker containers to docker hub? Or build every time inside the CI VM ?

**Comment by david-allison-1** [comment]
> Should we push the docker containers to docker hub? Or build every time inside the CI VM ?

Let's see how much more time it adds to the build. For now I'd be happy with things in code (possibly with a PERF: todo note) as it's one less service that we'll need to keep an eye on, and one less thing that can go wrong. 

Once things are stable, and we've had a release or two, I feel it'll be best to reconsider.

**Comment by david-allison-1** [comment]
```
error: no container engine found; install docker or podman
note: run with `RUST_BACKTRACE=1` for a backtrace
```

**Comment by TarekkMA** [comment]
Yeah forgot to install docker and build the images 😅

File: rsdroid/src/main/java/net/ankiweb/rsdroid/BackendUtils.java [hunk 8]
```diff
@@ -26,4 +28,9 @@
     public static void openAnkiDroidCollection(BackendV1 backendV1, String path) {
         backendV1.openAnkiDroidCollection(Backend.OpenCollectionIn.newBuilder().setCollectionPath(path).build());
     }
+
+    @NonNull
+    public static String getAnkiCommitHash() {
```
Comment (david-allison-1) [inline rsdroid/src/main/java/net/ankiweb/rsdroid/BackendUtils.java:33]: Think we should split this into a separate PR?

Comment (TarekkMA) [reply rsdroid/src/main/java/net/ankiweb/rsdroid/BackendUtils.java:33]: yes

**Comment by TarekkMA** [comment]
GH action didn't run on the previous 2 commits. Had to push an extra commit to make it work.

**Comment by TarekkMA** [comment]
@david-allison-1 we got a stuck one https://github.com/david-allison-1/Anki-Android-Backend/pull/39/checks?check_run_id=2244157423

**Comment by david-allison-1** [comment]
Restarted

**Comment by TarekkMA** [comment]
Is that a new thing on github?
Screen Shot 2021-05-07 at 12 01 08 AM

**Comment by david-allison-1** [comment]
> Is that a new thing on github?
> Screen Shot 2021-05-07 at 12 01 08 AM

Yeah, new feature: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

I thought I'd only need to approve once, but it seems it happens each time you push

Could you stick in a dummy commit somewhere so I can approve it?

**Comment by TarekkMA** [comment]
done

**Comment by TarekkMA** [comment]
I have cleaned the commit history and reworded/reordered some commits. let me know if we need to change anything extra.

the `doctor.sh` and `ENVIRONMENT.md` files might need some extra work to be done on other PRs.

File: .github/workflows/macos_build.yml [hunk 9]
```diff
@@ -89,7 +86,7 @@ jobs:
 
     # COULD_BE_BETTER: This may not be needed - tiny speed penalty
     - name: Init i18n submodule
-      run: git submodule update --init --recursive
+      run:  git submodule update --init --recursive --remote --force
```
Comment (david-allison-1) [inline .github/workflows/macos_build.yml:89]: can `submodules: 'true'` be removed above?

**Review by david-allison-1 — COMMENTED** [review]
Just one comment - happy to merge if CI is green, thanks so much!!

File: .github/workflows/macos_build.yml (hunk 9 above)
Comment (TarekkMA) [reply .github/workflows/macos_build.yml:89]: yes, it can.

Comment (TarekkMA) [reply .github/workflows/macos_build.yml:87]: done

**Comment by TarekkMA** [comment]
@david-allison-1 finally all green

**Review by david-allison-1 — APPROVED** [review]
You have no idea how much I appreciate this, thank you so much!

**Comment by TarekkMA** [comment]
It was a pleasure working on this PR, I learned about rust for the first time, and played with docker and GitHub actions. Thank you

**Comment by mikehardy** [comment]
Once @TarekkMA is a contributor in the repo (vs merged with no comments :laughing: ) further PRs will run without approval based on my experience in other repos
